src/context/walletContext/walletContext.tsx

```diff
@@ -51,166 +43,137 @@ export const WalletProvider = ({ children }: { children: ReactNode }) => {
                         walletProviderType: account.account_type as unknown as interfaces.WalletProviderType,
                         protocolType: interfaces.WalletProtocolType.EVM,
                         address: account.account_identifier,
-                        connected: false,
-                        balance: BigInt(0),
                     }
-                    if (isPeanut(wallet)) {
-                        wallet.balance = await peanutPublicClient.readContract({
+
+                    let balance = BigInt(0)
+                    if (isPeanut(dbWallet)) {
+                        balance = await peanutPublicClient.readContract({
                             address: USDC_ARBITRUM_ADDRESS,
                             abi: erc20Abi,
                             functionName: 'balanceOf',
-                            args: [getAddress(wallet.address)],
+                            args: [getAddress(dbWallet.address)],
                         })
                     }
+
+                    const wallet: interfaces.IWallet = {
+                        ...dbWallet,
+                        balance,
+                        connected: false, // Will be set in processedWallets memo
+                    }
+
                     return wallet
                 })
             return Promise.all(processedWallets)
         },
     })
 
-    const updateSelectedWalletWithConnectionStatus = useCallback(
-        (wallet: interfaces.IWallet) => {
-            if (isPeanut(wallet)) {
-                wallet.connected = isKernelClientReady && kernelClientAddress == wallet.address
-            } else {
-                wallet.connected = isWagmiConnected && wagmiAddress == wallet.address
-            }
-            setSelectedWallet(wallet)
-        },
-        [isKernelClientReady, kernelClientAddress, isWagmiConnected, wagmiAddress]
-    )
+    const selectedWallet = useMemo(() => {
+        if (!selectedAddress || !wallets.length) return undefined
+        const wallet = wallets.find((w) => w.address === selectedAddress)
+        if (!wallet) return undefined
 
-    /**
-     * Update initial selected wallet
-     */
-    useEffect(() => {
-        if (!selectedWallet && wallets && wallets.length > 0) {
-            const wallet = wallets.find(isPeanut) ?? wallets[0]
-            updateSelectedWalletWithConnectionStatus(wallet)
+        return {
+            ...wallet,
+            connected: isWalletConnected(wallet),
         }
-    }, [wallets, selectedWallet])
+    }, [selectedAddress, wallets, isWalletConnected])
 
-    /**
-     * Update selected wallet with connection status when kernel client is ready or wagmi is connected
-     */
+    // Set initial selected address
     useEffect(() => {
-        if (selectedWallet) {
-            updateSelectedWalletWithConnectionStatus(selectedWallet)
+        if (!selectedAddress && wallets.length > 0) {
+            const initialWallet = wallets.find(isPeanut) ?? wallets[0]
```

Comment:
Q: its smart to use peanut wallet as the default wallet, i like it. Question, do we respect users preferences? Ie should we store their last used wallet and let them use that instead on the next session?

Reply:
Hmmm I think that's a good Idea